### PR TITLE
Simplify IDeliveryClient and get rid of its internal dependencies/references

### LINKED ISSUE #87: Spike: Simplify IDeliveryClient and get rid of its internal dependencies/references
### Motivation

- `IDeliveryClient` contains members such as `IContentLinkUrlResolver`, `ICodeFirstModelProvider`, or `IInlineContentItemsProcessor` which shouldn't be part of the contract at all. Currently, they're there just because of internal referencing of the `DeliveryClient` in the underlying code.
- The wrong code can be easily identified:
  - All references to `this` in the `DeliveryClient` (and references to `IDeliveryClient` in the underlying code 
    - such as `_contentLinkResolver = new ContentLinkResolver(_client.ContentLinkUrlResolver);` in the  `ContentItem`.
    - or `_client.CodeFirstModelProvider.GetContentItemModel<T>(source, _response["modular_content"]))` in `DeliveryItemListingResponse<T>`

### Expected spike outcome

We should come up with a way to get rid of these dependencies and simplify the `IDeliveryClient` contract.

Things to consider:
- we could utilize DI, however we'd like to avoid using any particular DI framework
  - we don't want to push consumers of the library to use a specific DI framework
  - we might want to introduce some simple DI of our own making, crafted just to fit our needs
    - there would be a default implementation of the container with a possibility to replace the logic
    - it should also be possible to inject a DI framework of user's choice to take care of the instantiation completely
      - is there some kind of generic DI interface that we could expose for replacement?
  - we should keep in mind that the user can create multiple instances of the `DeliveryClient` in a single application. So a **simple** singleton might not be the right thing for us.
- another approach would be to refactor the logic in a way that no DI is needed. 
  - in a way that all the complex stuff happens outside of the models/response objects and the models/responses are just dummy container objects

### PR DESCRIPTION
**Pull request by matus12**
### Motivation

Which issue does this fix? Fixes #87 

### Checklist

- [x] Code follows coding conventions held in this repo
- [x] Automated tests have been added
- [x] Tests are passing
- [x] Docu has been updated (if applicable)
- [x] Temporary settings (e.g. project ID used during development and testing) have been reverted to defaults

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -102,12 +117,12 @@ See [Working with Strongly Typed Models](https://github.com/Kentico/delivery-sdk
...
+To retrieve unpublished content, you need to create a `DeliveryClient` with both Project ID and Preview API key. Each Kentico Cloud project has its own Preview API key.
 
 ```csharp
 // Note: Within a single project, we recommend that you work with only
 // either the production or preview Delivery API, not both.
-DeliveryClient client = new DeliveryClient("YOUR_PROJECT_ID", "YOUR_PREVIEW_API_KEY");
+DeliveryClient client = DeliveryClientBuilder.WithOptions(builder => builder.WithProjectId("YOUR_PROJECT_ID").UsePreviewApi("YOUR_PREVIEW_API_KEY"));
 ```
 
 For more details, see [Previewing unpublished content using the Delivery API](https://developer.kenticocloud.com/docs/preview-content-via-api).
```
[the remark this answers is not in the document]
Comment (matus12) [reply README.md]: Docs excluding GH are being modified by @jancerman

**Comment by matus12** [comment]
**Breaking changes:** 

- Removed .NET Framework 4.5 as a target
- class `DeliveryClient` is no longer public - requests are now made through instance implementing `IDeliveryClient` which can be created using new `DeliveryClientBuilder` class
- Added a builder class for `DeliveryOptions`
- Added an extension method on `IServiceCollection` that registers `IDeliveryClient` implementation
- Custom implementation of resolvers, processors, mappers can no longer be set to public properties - now they can be set through `DeliveryClientBuilder` class or by registering them to the `ServiceCollection`

File: KenticoCloud.Delivery/Builders/DeliveryOptions/DeliveryOptionsBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                _deliveryOptions.EnableResilienceLogic = true;
+
+                return this;
+            }
+        }
+
+        IOptionsSteps IOptionsSteps.WithMaxRetryAttempts(int attempts)
+        {
+            DeliveryOptionsValidator.ValidateMaxRetryAttempts(attempts);
+            _deliveryOptions.MaxRetryAttempts = attempts;
```
[the remark this answers is not in the document]
Comment (matus12) [reply KenticoCloud.Delivery/Builders/DeliveryOptions/DeliveryOptionsBuilder.cs]: You are right, I was relying on the fact that it is true by default, I will add it there.

File: KenticoCloud.Delivery/Builders/DeliveryOptions/IDeliveryOptionsBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+        /// <summary>
+        /// An optional step that sets the client to wait for updated content. It should be used when you are acting upon a webhook call.
+        /// </summary>
+        IOptionsSteps WaitForLoadingNewContent { get; }
+
+        /// <summary>
+        /// An optional step that sets the maximum number of retry attempts.
+        /// </summary>
+        /// <param name="attempts">Maximum retry attempts</param>
```
[the remark this answers is not in the document]
Comment (matus12) [reply KenticoCloud.Delivery/Builders/DeliveryOptions/IDeliveryOptionsBuilder.cs]: Hm, maybe I'm missing something but I'm not sure about the third point, using custom `IResiliencePolicyProvider` and obtaining the value. Either I hardcode the value into my custom provider or I set it to `DeliveryOptions` (builder/config) so I know I have it there and I can retrieve it, right?

File: KenticoCloud.Delivery/Builders/DeliveryOptions/IDeliveryOptionsBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        IOptionsSteps WithMaxRetryAttempts(int attempts);
+
+        /// <summary>
+        /// An optional step that sets a custom endpoint for a chosen API.
+        /// </summary>
+        /// <param name="customEndpoint">A custom endpoint address.</param>
+        IOptionsSteps WithCustomEndpoint(string customEndpoint);
```
[the remark this answers is not in the document]
Comment (matus12) [reply KenticoCloud.Delivery/Builders/DeliveryOptions/IDeliveryOptionsBuilder.cs]: If we allowed `Uri` too, we would check if the Uri is absolute, if not, throw an exception, right? Then validate it and store it in the options as string? 😄

File: KenticoCloud.Delivery/Builders/DeliveryClient/DeliveryClientBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        public IDeliveryClientOptionalSteps BuildWithProjectId(string projectId)
+            => BuildWithDeliveryOptions(builder => builder.WithProjectId(projectId).UseProductionApi.Build());
+
+        public IDeliveryClientOptionalSteps BuildWithProjectId(Guid projectId)
+            => BuildWithDeliveryOptions(builder => builder.WithProjectId(projectId).UseProductionApi.Build());
+
+        IDeliveryClientOptionalSteps IDeliveryClientOptionalSteps.WithHttpClient(HttpClient httpClient)
+        {
+            _httpClient = httpClient ??
+                          throw new ArgumentNullException(nameof(httpClient), "Http client is not specified");
```
[the remark this answers is not in the document]
Comment (matus12) [reply KenticoCloud.Delivery/Builders/DeliveryClient/DeliveryClientBuilder.cs]: Looks like my Resharper is not configured properly. 😭

File: KenticoCloud.Delivery.Tests/DeliveryClientTests.cs [hunk 6]
```diff
@@ -15,15 +15,20 @@ namespace KenticoCloud.Delivery.Tests
 {
     public class DeliveryClientTests
```
Comment (vaclavm) [inline KenticoCloud.Delivery.Tests/DeliveryClientTests.cs]: Format the whole file, so it has correct formatting from up to bottom.

**Pull request merged**